[DDP] Point grads to bucket if None and grad_as_bucket_view.

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #96439

Per title

Gating change for now to test internally.

Differential Revision: [D43860503](https://our.internmc.facebook.com/intern/diff/D43860503/)

rohan-varma committed:
[DDP] Point grads to bucket if None and grad_as_bucket_view.

Per title

Differential Revision: [D43860503](https://our.internmc.facebook.com/intern/diff/D43860503/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Point grads to bucket if None and grad_as_bucket_view."


Per title

Differential Revision: [D43860503](https://our.internmc.facebook.com/intern/diff/D43860503/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Point grads to bucket if None and grad_as_bucket_view."


Per title

Differential Revision: [D43860503](https://our.internmc.facebook.com/intern/diff/D43860503/)

[ghstack-poisoned]

zhaojuanmao commented on torch/nn/parallel/distributed.py:
```diff
@@ -1344,6 +1344,11 @@ def forward(self, *inputs, **kwargs):
                 assert self.logger is not None
                 self.logger.set_runtime_stats_and_log()
                 self.num_iterations += 1
+                if self.num_iterations > 1 and any(
+                    p.grad is None for p in self.module.parameters()
+                ):
+                    self.reducer._point_grads_to_bucket()
```
wondering whether it will be better to move to the end of forward, so that it is close to beginning of backward just in case:)

rohan-varma committed:
Update on "[DDP] Point grads to bucket if None and grad_as_bucket_view."


Per title

Differential Revision: [D43860503](https://our.internmc.facebook.com/intern/diff/D43860503/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Point grads to bucket if None and grad_as_bucket_view."


Per title

Differential Revision: [D43860503](https://our.internmc.facebook.com/intern/diff/D43860503/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Point grads to bucket if None and grad_as_bucket_view."


Per title

Differential Revision: [D43860503](https://our.internmc.facebook.com/intern/diff/D43860503/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Point grads to bucket if None and grad_as_bucket_view."


Per title

Differential Revision: [D43860503](https://our.internmc.facebook.com/intern/diff/D43860503/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Point grads to bucket if None and grad_as_bucket_view."


Per title

Differential Revision: [D43860503](https://our.internmc.facebook.com/intern/diff/D43860503/)

[ghstack-poisoned]

awgu commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -336,6 +336,33 @@ void Reducer::check_grad_layout(
+  std::unordered_set<int> zeroed_buckets;
+  for (size_t variable_index = 0; variable_index < params_.size();
+       ++variable_index) {
+    const auto& bucket_index = variable_locators_[variable_index];
```
nit:
```suggestion
    const auto& variable_locator = variable_locators_[variable_index];
```

`variable_locator.bucket_index` may be easier to understand than `bucket_index.bucket_index`, which may be a misnomer.

awgu commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -336,6 +336,33 @@ void Reducer::check_grad_layout(
 }
 
+void Reducer::point_grads_to_bucket() {
+  std::unordered_set<int> zeroed_buckets;
```
nit:
```suggestion
  std::unordered_set<int> zeroed_bucket_indices;
```

awgu commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -336,6 +336,33 @@ void Reducer::check_grad_layout(
+      if (!grad.defined()) {
+        // grad is None: assuming user called zero_grad(set_to_none=True)
+        // point grad to bucket
+        grad = bucket_view;
```
C++ noob question: Will the `bucket_view` reference be bound to the anonymous grad callback's closure automatically (even though the callback is only defined to take in the `grad` reference)?

awgu replied:
Also, does assigning a reference to a reference persist? Are we sure that the `grad` is being set to `bucket_view` even after this lambda returns?

rohan-varma replied:
In the lambda, the stuff in `[...]` defines the variables to capture, and `[&]` means to capture the surrounding context by reference. `(auto& grad)` is the argument taken into this callback.

I believe this will work as we do the same thing here: https://fburl.com/code/pshiibxd,

awgu reviewed (approved):
LGTM! Lint error looks real.
